docs/proposals/20210222-machineadm.md

```diff
@@ -0,0 +1,835 @@
+}
+```
+#### Bootstrap template ConfigMap/Secret format:
+
+Bootstrap templates will be simple go template strings that will form the final
+user data block. For the purposes of machineadm, the OS bootstrapper must write
+the file to disk somewhere, and execute machineadm with administrative credentials
+against the provided file.
+
+The following variables and functions are supported:
+
+##### Variables
+
+* `machine_config`: The contents of the final infrastructure customized
+  bootstrap configuration document.
+
+##### Custom supported functions
+* `base64`: Base64 encodes the input.
+* `gzipBase64`: gzip compresses the input and then base64 encodes it.
+
+An example using cloud-init is presented below:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: cloud-init-machineadm-bootstrap-template
+data:
+  template: |
+    #cloud-config
+    write_files:
+    - encoding: gz+b64
+      content:  {{ machine_config | gzipBase64 }}.
+      owner: root:root
+      path: /tmp/machineadm_config.yaml
+      permissions: '0644'
+    runcmd:
+      - /usr/bin/machineadm --bootstrap --path /tmp/machineadm_config.yaml
+```
+#### MachineConfig types
+
+It is important to note that we have in effect two representations of the MachineConfig: First, we have the
+MachineBootstrap Kubernetes resource that may reference other Kubernetes resources such as secrets for inclusion
+in the machine configuration. Secondly, we have the "serialized" MachineConfiguration which is the "fully realized"
+configuration, that inlines all of the referenced data and is stored as its own Kubernetes secret for handoff to
+the infrastructure provider for secure storage.
+
+##### Common types
+
+```go
+// MachineBootstrapConfig is a resource that defines the configuration
+// of a machine
+type MachineBootstrapConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MachineBootstrapConfigSpec   `json:"spec,omitempty"`
+	Status MachineBootstrapConfigStatus `json:"status,omitempty"`
+}
+
+// MachineBootstrapConfigSpec defines the desired state of a MachineBootstrapConfig
+type MachineBootstrapConfigSpec struct {
+	PluginRefs []corev1.ObjectReference `json:"pluginRefs"`
+}
+
+// MachineBootstrapConfigTemplate is a resource that defines the configuration
+// of a machine
+type MachineBootstrapConfigTemplate struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec MachineBootstrapConfigTemplateSpec `json:"spec,omitempty"`
+}
+
+// MachineBootstrapConfigTemplateSpec defines the desired state of a MachineBootstrapConfigTemplate
+type MachineBootstrapConfigTemplateSpec struct {
+	MachineBootstrapConfigSpec `json:",inline"`
+	// PluginTemplateRefs is a list of machine bootstrap plugin template refs
+	// which will be instantiated for each machine.
+	PluginTemplateRefs []corev1.ObjectReference `json:"pluginTemplateRefs,omitempty"`
```

Comment:
Might be I'm missing something, but what is the difference between `PluginTemplateRefs`  and `PluginRefs` embedded via `MachineBootstrapConfigSpec`?